### ISSUE TITLE
Add support/migrate to MV3

### ISSUE THREAD

**Issue by LLCampos** [AlexITC/chrome-scalajs-template#28]
[Manifest V3 for Chrome Extensions ](https://developer.chrome.com/docs/extensions/mv3/intro/) was introduced recently.

Currently the code only supports MV2.

**Comment by AlexITC** [comment]
What's the state for the Manifest V3 in Firefox? I don't plan to do such migration unless it's supported by Firefox (I couldn't find their current state), also, this should be implemented in the plugin (https://github.com/AlexITC/scala-js-chrome) first.

Feel free to raise a PR.

**Comment by phdoerfler** [comment]
Little known fact: Safari has also added support for Manifest v2 extensions. Regarding support for MV3: A quick search brought up this [12 months old forum post](https://developer.apple.com/forums/thread/650463) stating:
> Safari 14 only supports manifest v2. Manifest v3 is still in development, and it is not currently shipping in any browser. We can't comment on future Safari releases.

**Comment by AlexITC** [comment]
Apparently, this needs to get done soon ([see](https://developer.chrome.com/blog/mv2-transition/)), first thing is to get it done in github.com/AlexITC/scala-js-chrome

**Comment by phdoerfler** [comment]
Indeed. In particular:
> **January 17, 2022**: New Manifest V2 extensions will no longer be accepted by the Chrome Web Store. Developers may still push updates to existing Manifest V2 extensions, but no new Manifest V2 items may be submitted.

**Comment by AlexITC** [comment]
Indeed, that makes the template/plugin not very useful because people won't be able to publish to the chrome web store.

I'm yet to find what exactly needs to change, hoping that it isn't much work.

I'll keep the ticket opened to post updates.

**Comment by phdoerfler** [comment]
I gingerly prodded at the generated manifest to see how quickly it falls apart when changing `"manifest_version": 2,` to `"manifest_version": 3,`:
Aside from this, a major, necessary change is this:
https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/

Where `scripts` supported multiple scripts (which the template makes use of), `service_worker` only supports one single .js file. Currently these scripts are:
- scripts/common.js (just holds one small facade that makes it easier to interact with some JS from Scala)
- main-bundle.js (the main event)
- scripts/background-script.js (just has a single line `runOnBackground();`)

These would need to be fused into one, e.g., by merging them all into `main-bundle.js` or by having one hand written js file containing the code of common.js and background-script.js and calling main-bundle.js, if that is possible.
Lastly, there's the hacky way of generating a js file that simply concatenates / sensibly merges all three of these js files using an sbt task.

But the migration to a service worker isn't done here, as the docs point out:
> First, service workers are terminated when not in use and restarted when needed (similar to event pages). Second, service workers don't have access to DOM.
The docs also link to how these two issues can be addressed

**Comment by AlexITC** [comment]
@phdoerfler thanks for sharing the issues you have found, the migration to service workers is the major change I have found.

As you point out, we can trick the build to produce a single js.

I have been waiting to see how [Metamask](https://github.com/MetaMask/metamask-extension/issues/13411) handles this migration.

**Comment by phdoerfler** [comment]
I see, thanks for the update! It doesn't seem like Metamask is willing to share their findings, no? It might be that we have to rely on Chrome's documentation.

Speaking of:
I have concatenated all three files, just to see what Chrome is thinking of them, now. The next construction site was thus revealed:

> Error at key 'web_accessible_resources'. Parsing array failed at index 0: expected dictionary, got string

Continuing in this Frankenstein spirit, replacing:

```
"web_accessible_resources": [
  "icons/*",
  "scripts/common.js",
  "main-bundle.js",
  "scripts/active-tab-website-script.js"
]
```

with:

```
"web_accessible_resources": [{
  "resources": [
    "icons/*",
    "scripts/common.js",
    "main-bundle.js",
    "scripts/active-tab-website-script.js"
  ],
  "matches": ["https://www.example.com/*", "https://www.example.de/*"]
}],
```

got Chrome to accept the manifest. Victory! Or is it?
Unsurprisingly, the extension fails to load because of what I tried to sell Chrome as a "service worker":
<img width="644" alt="image" src="https://user-images.githubusercontent.com/163428/165327073-2176854c-b7fa-47b1-93ec-2277a675b755.png">

FWIW, you probably came to similar findings yourself already. Just thought I'd share what doesn't work so the next one does not have to repeat these mistakes but can instead make new ones 😉

**Comment by AlexITC** [comment]
> I see, thanks for the update! It doesn't seem like Metamask is willing to share their findings, no? It might be that we have to rely on Chrome's documentation and experimentation.

If anyone wants to upgrade soon, for sure, we should rely on the public docs + trial and error, in my personal case, I do not have this as a blocker which is why I have been waiting until Metamask releases the support for MV3, even if they do not share their analysis, the code should get in their repo some day.

>  Unsurprisingly, the extension fails to load because of what I tried to sell Chrome as a "service worker":

I do not remember much about this but I understand that the service worker api for websites has slightly different apis available for extensions (which could differ from what background context has in MV2), which could explain problems like the one you are facing.

A quick search makes me believe that we could import many js files for a service worker:
- https://stackoverflow.com/questions/66406672/chrome-extension-mv3-modularize-service-worker-js-file
- https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/

**Comment by phdoerfler** [comment]
Any news on this?

**Comment by AlexITC** [comment]
Unfortunately, I haven't had a chance to look into this, right now I'm not actively working on any browser extension.
